### ISSUE TITLE
SQL and virtual hosts

### ISSUE THREAD

**Issue by ednt** [processone/docs.ejabberd.im#126]
We have more virtual hosts on one ejabberd server.
Since it is written to use an external database if the usage/load gets higher, we decided to use mysql/mariadb.

Following the 'official' way https://www.process-one.net/blog/install-and-configure-mariadb-with-ejabberd/
I noticed very fast, that this is not working for more then one virtual host.
The schema can not decide between them.

After digging arround I found the mysql.new.sql and I was lucky ... in the first moments.
But then I noticed that this is, up to now, not the official way.

So what should I do with my 21 (and growing) virtual hosts???

Using the 'unofficial' mysql.new.sql schema or different databases for each virtual host?

Btw. I use version 23.01-1 because it is the default package in debian 12 bookworm.

The documentation at https://docs.ejabberd.im/tutorials/mysql/
also shows the 'old' way and it is not explained how to use an own database per virtual host.
The new schema is not mentioned. Should we use it or not?

Please let the users know how to do it.

**Comment by licaon-kter** [comment]
-backports has 23.10 fyi https://packages.debian.org/bookworm-backports/ejabberd ( see upgrade notes: https://docs.ejabberd.im/admin/upgrade/#specific-version-upgrade-notes )

no, that won't fix your issue :)

if you don't use .new.sql yes, you need more databases

**Comment by weiss** [comment]
As long as we support both schemas, I guess it would help to document things clearly within the [Using ejabberd with MySQL][1] tutorial, e.g. by referencing the [existing text][2].

[1]: https://docs.ejabberd.im/tutorials/mysql/
[2]: https://docs.ejabberd.im/admin/configuration/database/#default-and-new-schemas

**Comment by badlop** [comment]
I guess by now you have already followed the [ejabberd Documentation site](https://docs.ejabberd.im/) -> For Administrators -> Configuration -> Databases -> and there, taking a look at the different sections, you finally found [Default and New Schemas](https://docs.ejabberd.im/admin/configuration/database/#default-and-new-schemas) which clarifies most of your doubts. Let's review them:

> Since it is written to use an external database if the usage/load gets higher, we decided to use mysql/mariadb.

Right. The default internal Mnesia database is suitable only for small servers. For medium or larger ones better to use a SQL database.

> We have more virtual hosts on one ejabberd server.

If it is just a few vhosts with millions of accounts, you may want to use the default schema, setting a different database for each vhost. It requires more configuration steps, and saves memory because each table in each database does not have to store the hostname.

If they are a lot of vhosts which may be small or medium or large, then you may prefer to use the new schema in one single database. This saves a lot of configuration time.

> Following the 'official' way https://www.process-one.net/blog/install-and-configure-mariadb-with-ejabberd/

That is a tutorial. It was written in 2021, for ejabberd 21.01. When software improves, the old tutorials may get obsolete. That is not documentation.

Documentation is written one day, and gets updated everytime to make sure it is always up-to-date.


> After digging arround I found the mysql.new.sql and I was lucky ... in the first moments. But then I noticed that this is, up to now, not the official way.

Not official? Where did you get that impression?

The new schemas for all the databases are documented in https://docs.ejabberd.im/admin/configuration/database/#default-and-new-schemas and the files are in https://github.com/processone/ejabberd/tree/master/sql

> So what should I do with my 21 (and growing) virtual hosts???
> Using the 'unofficial' mysql.new.sql schema or different databases for each virtual host?

Using `mysql.new.sql`

> The documentation at https://docs.ejabberd.im/tutorials/mysql/ also shows the 'old' way and it is not explained how to use an own database per virtual host. The new schema is not mentioned. Should we use it or not?

That is a tutorial included in the documentation site. As you noticed, the tutorial at https://docs.ejabberd.im/tutorials/mysql/ was obsolete compared to the documentation in https://docs.ejabberd.im/admin/configuration/database/ and the source code available in git.

A tutorial is not reference documentation. A tutorial describes one method to do something, either the easiest, or the most popular, or the most common, or a new surprising method, or the method preferred by the tutorial author. Following a tutorial is a great way to start learning how to use a feature at a basic level. Once the tutorial is finished and you get the software running, the next thing to do is read the documentation to find all the other possibilities that the tutorial did not explain.

A tutorial only covers a feature lightly, the reference documentation covers the feature deep and long.

> So what should I do with my 21 (and growing) virtual hosts???
> Please let the users know how to do it.

With many small or medium vhosts, the new schema is preferable.

I've updated the [MySQL tutorial](https://docs.ejabberd.im/tutorials/mysql/) to point readers of the tutorial to the default/new schema possibility, and to the new automatic schema update feature.
